src/mds/MDSPerfMetricTypes.h

```diff
@@ -0,0 +1,189 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef CEPH_MDS_PERF_METRIC_TYPES_H
+#define CEPH_MDS_PERF_METRIC_TYPES_H
+
+#include <iostream>
+
+#include "include/denc.h"
+#include "include/utime.h"
+#include "mdstypes.h"
+
+enum UpdateType {
+  UPDATE_TYPE_REFRESH = 0,
+  UPDATE_TYPE_REMOVE,
+};
+
+struct CapHitMetric {
+  uint64_t hits = 0;
+  uint64_t misses = 0;
+
+  DENC(CapHitMetric, v, p) {
+    DENC_START(1, 1, p);
+    denc(v.hits, p);
+    denc(v.misses, p);
+    DENC_FINISH(p);
+  }
+
+  friend std::ostream& operator<<(std::ostream& os, const CapHitMetric &metric) {
+    os << "{hits=" << metric.hits << ", misses=" << metric.misses << "}";
+    return os;
+  }
+};
+
+struct ReadLatencyMetric {
+  utime_t lat;
+
+  DENC(ReadLatencyMetric, v, p) {
+    DENC_START(1, 1, p);
+    denc(v.lat, p);
+    DENC_FINISH(p);
+  }
+
+  friend std::ostream& operator<<(std::ostream& os, const ReadLatencyMetric &metric) {
+    os << "{latency=" << metric.lat << "}";
+    return os;
+  }
+};
+
+struct WriteLatencyMetric {
+  utime_t lat;
+
+  DENC(WriteLatencyMetric, v, p) {
+    DENC_START(1, 1, p);
+    denc(v.lat, p);
+    DENC_FINISH(p);
+  }
+
+  friend std::ostream& operator<<(std::ostream& os, const WriteLatencyMetric &metric) {
+    os << "{latency=" << metric.lat << "}";
+    return os;
+  }
+};
+
+struct MetadataLatencyMetric {
+  utime_t lat;
+
+  DENC(MetadataLatencyMetric, v, p) {
+    DENC_START(1, 1, p);
+    denc(v.lat, p);
+    DENC_FINISH(p);
+  }
+
+  friend std::ostream& operator<<(std::ostream& os, const MetadataLatencyMetric &metric) {
+    os << "{latency=" << metric.lat << "}";
+    return os;
+  }
+};
+
+WRITE_CLASS_DENC(CapHitMetric)
+WRITE_CLASS_DENC(ReadLatencyMetric)
+WRITE_CLASS_DENC(WriteLatencyMetric)
+WRITE_CLASS_DENC(MetadataLatencyMetric)
+
+// metrics that are forwarded to the MDS by client(s).
+struct Metrics {
+  // metrics
+  CapHitMetric cap_hit_metric;
+  ReadLatencyMetric read_latency_metric;
+  WriteLatencyMetric write_latency_metric;
+  MetadataLatencyMetric metadata_latency_metric;
+
+  // metric update type
+  UpdateType update_type = UpdateType::UPDATE_TYPE_REFRESH;
+
+  void encode(bufferlist &bl) const {
+    using ceph::encode;
+
+    ENCODE_START(1, 1, bl);
+    encode(static_cast<uint32_t>(update_type), bl);
```

Comment:
no need to cast the enum to the underlying type if it has an underlying type like
```c++
enum UpdateType : uint32_t
```